File: .claude/skills/afip-arca-wsaa/examples/python/wsaa_login.py
"""
WSAA login — get a TA for a given service (homologación by default).

Usage:
    python wsaa_login.py --cert path/to/cert.crt --key path/to/private.key --service wsfe

Outputs the TA XML to stdout. Save to a file and cache it until <expirationTime>.
"""
from __future__ import annotations

import argparse
import base64
import sys
from datetime import datetime, timedelta, timezone
from pathlib import Path

from cryptography import x509
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.serialization import pkcs7
from zeep import Client

WSAA_HOMO = "https://wsaahomo.afip.gov.ar/ws/services/LoginCms?wsdl"
WSAA_PROD = "https://wsaa.afip.gov.ar/ws/services/LoginCms?wsdl"

ART = timezone(timedelta(hours=-3))


def build_tra(service: str) -> bytes:
    now = datetime.now(ART)
    tra = f"""<?xml version="1.0" encoding="UTF-8"?>
<loginTicketRequest version="1.0">
  <header>
    <uniqueId>{int(now.timestamp())}</uniqueId>
    <generationTime>{(now - timedelta(minutes=1)).isoformat(timespec='seconds')}</generationTime>
    <expirationTime>{(now + timedelta(minutes=10)).isoformat(timespec='seconds')}</expirationTime>
  </header>
  <service>{service}</service>
</loginTicketRequest>
"""
    return tra.encode("utf-8")


def sign_cms(tra_bytes: bytes, cert_path: Path, key_path: Path) -> bytes:
    cert = x509.load_pem_x509_certificate(cert_path.read_bytes())
    key = serialization.load_pem_private_key(key_path.read_bytes(), password=None)
    cms = (
        pkcs7.PKCS7SignatureBuilder()
        .set_data(tra_bytes)
        .add_signer(cert, key, hashes.SHA256())
        .sign(serialization.Encoding.DER, [pkcs7.PKCS7Options.Binary])
    )
    return cms


def login(cms_b64: str, env: str) -> str:
    wsdl = WSAA_HOMO if env == "homo" else WSAA_PROD
    client = Client(wsdl=wsdl)
    return client.service.loginCms(in0=cms_b64)


def main() -> int:
    p = argparse.ArgumentParser(description="WSAA login — obtain a TA")
    p.add_argument("--cert", type=Path, required=True)
    p.add_argument("--key", type=Path, required=True)
    p.add_argument("--service", default="wsfe", help="WS scope: wsfe, wsfex, ws_sr_padron_a4, ...")
    p.add_argument("--env", choices=["homo", "prod"], default="homo")
    args = p.parse_args()

    tra = build_tra(args.service)
    cms = sign_cms(tra, args.cert, args.key)
    cms_b64 = base64.b64encode(cms).decode("ascii")
    ta_xml = login(cms_b64, args.env)
    print(ta_xml)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: .claude/skills/afip-arca-wsaa/examples/node/wsaa-login.ts
/**
 * WSAA login — TypeScript example.
 * Usage: npx tsx wsaa-login.ts --cert cert.crt --key key.key --service wsfe [--env homo|prod]
 */
import { readFileSync } from "node:fs";
import { parseArgs } from "node:util";
import forge from "node-forge";
import * as soap from "soap";

const WSAA_HOMO = "https://wsaahomo.afip.gov.ar/ws/services/LoginCms?wsdl";
const WSAA_PROD = "https://wsaa.afip.gov.ar/ws/services/LoginCms?wsdl";

function isoArt(ts: number): string {
  // Format unix seconds as ISO 8601 with -03:00 (ART) offset.
  const d = new Date((ts - 3 * 3600) * 1000);
  const pad = (n: number) => String(n).padStart(2, "0");
  return (
    `${d.getUTCFullYear()}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())}` +
    `T${pad(d.getUTCHours())}:${pad(d.getUTCMinutes())}:${pad(d.getUTCSeconds())}-03:00`
  );
}

function buildTra(service: string): string {
  const now = Math.floor(Date.now() / 1000);
  return `<?xml version="1.0" encoding="UTF-8"?>
<loginTicketRequest version="1.0">
  <header>
    <uniqueId>${now}</uniqueId>
    <generationTime>${isoArt(now - 60)}</generationTime>
    <expirationTime>${isoArt(now + 600)}</expirationTime>
  </header>
  <service>${service}</service>
</loginTicketRequest>`;
}

function signCms(tra: string, certPem: string, keyPem: string): string {
  const cert = forge.pki.certificateFromPem(certPem);
  const key = forge.pki.privateKeyFromPem(keyPem);
  const p7 = forge.pkcs7.createSignedData();
  p7.content = forge.util.createBuffer(tra, "utf8");
  p7.addCertificate(cert);
  p7.addSigner({
    key,
    certificate: cert,
    digestAlgorithm: forge.pki.oids.sha256,
  });
  p7.sign({ detached: false });
  const der = forge.asn1.toDer(p7.toAsn1()).getBytes();
  return forge.util.encode64(der);
}

async function main(): Promise<void> {
  const { values } = parseArgs({
    options: {
      cert: { type: "string" },
      key: { type: "string" },
      service: { type: "string", default: "wsfe" },
      env: { type: "string", default: "homo" },
    },
  });
  if (!values.cert || !values.key) {
    throw new Error("--cert and --key are required");
  }

  const certPem = readFileSync(values.cert, "utf8");
  const keyPem = readFileSync(values.key, "utf8");
  const tra = buildTra(values.service!);
  const cmsB64 = signCms(tra, certPem, keyPem);

  const wsdl = values.env === "prod" ? WSAA_PROD : WSAA_HOMO;
  const client = await soap.createClientAsync(wsdl);
  const [result] = await client.loginCmsAsync({ in0: cmsB64 });
  process.stdout.write((result as { loginCmsReturn: string }).loginCmsReturn);
}

main().catch((e: unknown) => {
  console.error(e);
  process.exit(1);
});
